### PR TITLE
Support different-length pos/neg prompts for FLUX.1-schnell variants like Chroma

### PR DESCRIPTION
**Pull request by josephrocca**
# What does this PR do?

## Context:
[Chroma](https://huggingface.co/lodestones/Chroma) is a large-scale Apache 2.0 fine-tune of FLUX.1 Schnell. It is currently one of the top trending text-to-image models, and has been for several days now:

image

Someone recently asked about diffusers support:

* https://huggingface.co/lodestones/Chroma/discussions/5

I've currently got it working in diffusers:

* https://gist.github.com/josephrocca/385d9868ac52ea6f854b3ab96ec0ad25

but as you can see from the comments at the top of that script, it requires a couple of changes to diffusers source code for it to work out of the box.

## Changes:
One such change is due to Chroma requiring masking/truncation of prompts (all but the final padding token).

Currently diffusers requires that prompts are the same length, since it assumes that the full 512 T5 tokens will be used for both positive and negative prompts.

So `check_inputs` blocks it, and if we remove that check, then we get this error:
```
  File "/opt/conda/lib/python3.11/site-packages/diffusers/pipelines/flux/pipeline_flux.py", line 904, in __call__
    neg_noise_pred = self.transformer(
                     ^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/accelerate/hooks.py", line 176, in new_forward
    output = module._old_forward(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/diffusers/models/transformers/transformer_flux.py", line 522, in forward
    encoder_hidden_states, hidden_states = block(
                                           ^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/diffusers/models/transformers/transformer_flux.py", line 180, in forward
    attention_outputs = self.attn(
                        ^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/diffusers/models/attention_processor.py", line 588, in forward
    return self.processor(
           ^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/diffusers/models/attention_processor.py", line 2318, in __call__
    query = apply_rotary_emb(query, image_rotary_emb)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/diffusers/models/embeddings.py", line 1208, in apply_rotary_emb
    out = (x.float() * cos + x_rotated.float() * sin).to(x.dtype)
           ~~~~~~~~~~^~~~~
```

So we need to pass the negative prompt ids into the negative prompt forward pass, instead of passing the positive prompt ids into both.

## Who can review?

@yiyixuxu @sayakpaul @DN6

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your PR!

Before we get to reviewing the PR, could you please provide some side-by-side results with Schnell and Chroma on the same inputs (including the seeds)?

**Comment by asomoza** [comment]
there some more context in this issue: https://github.com/huggingface/diffusers/issues/11010

**Comment by josephrocca** [comment]
> could you please provide some side-by-side results with Schnell and Chroma on the same inputs (including the seeds)?

Oh, sure thing, see links below for some comparison grids - but some quick notes:
* If you're wondering how different/trained/diverged Chroma is from Schnell, these pics may be useful.
* If you're looking to understand how "high quality" or "aesthetic" Chroma is, then these will not be particularly useful, since Chroma is still training, and has almost entirely been trained on 512<sup>2</sup>px images so far. I think it's not ready to be compared on aesthetics or fine details at this point.
* I used ChatGPT 4.5 to generate the prompts used for these: https://chatgpt.com/share/67dc3cb0-fbb4-8007-a661-0184968418ad

## Image Grids:

* [Chroma](https://github.com/user-attachments/assets/98ed0c4f-d040-4b9b-a2d9-6557a260cce2)

* [Schnell](https://github.com/user-attachments/assets/cb578564-8a59-4e1c-af4a-6ae934746b08)

* [Chroma + aesthetic keywords](https://github.com/user-attachments/assets/9bcfbf37-f700-456c-a3c0-1e45742789d7) - a crude attempt to replicate aesthetic/preference tuning using some aesthetics-related keywords before and after the prompt. See the commented notes in [this gist](https://gist.github.com/josephrocca/385d9868ac52ea6f854b3ab96ec0ad25) for the prompt I used.

Also, if you skim the above grids, note that some of the images from Schnell look quite "clean" and coherent, and this is definitely an advantage that Schnell currently has, but note that in some cases Chroma should arguably win based on the style specified in the prompt ("courtroom sketch"):

image

Compared to Chroma:

image (1)

And Chroma with aesthetic keywords to try to emulate aesthetic tuning that Schnell has:

image (2)

You can see that although Schnell's is cleaner (and arguably slightly more coherent, though sample size is a bit small here), Chroma is definitely more faithful to the style specified in the prompt.

Also note that, as with other models that haven't had CFG baked in, you can get entirely different 'vibes' by tweaking Chroma's CFG - above I've used 5, with 20 steps (lodestone is [currently](https://github.com/lodestone-rock/simple_flow/commit/0c45449bd8038658cae7c9f7008f5ba61fea8b7a) working towards a few-step lora).

My experience so far with testing Chroma is that it has a lot more "soul" than Schnell and Dev - it's quite fun to play with.

**Comment by nitinmukesh** [comment]
Awesome. I looked at some of the outputs, Anubis one is amazing among many others.

**Comment by josephrocca** [comment]
Side note: Playing with the official Chroma ComfyUI workflow just now, I noticed that there are some potential differences in quality/coherence compared to my diffusers code which generated the above image - e.g. notice the alignment to the "bored human judge" in this seed=0 image with Chroma, which was less evident in the above examples:



So please take the above example images with a pinch of salt - Chroma quality may be better than what I've shown here. It could be due to quantization, or subtitles around ComfyUI sampling. I'd need to take bigger sample sizes to know if ComfyUI outputs are actually better, but I'm going to sleep now :)

**Review by hlky — APPROVED** [review]
Thanks @josephrocca

**Comment by nitinmukesh** [comment]
Just a quick query. After this is merged,  pad_tensors_to_equal_length will not be required anymore?

In my code I was facing issue when Prompt and N.P. are not of same length. So I had to add additional code to make them equal
https://github.com/newgenai79/sd-diffuser-webui/blob/main/modules/svdquant/tab_flux_dev.py#L157

pad_...
https://github.com/newgenai79/sd-diffuser-webui/blob/main/modules/util/utilities.py#L55

**Comment by josephrocca** [comment]
I'm not sure about the conventions in diffusers, but since prompt truncation is equivalent to prompt masking, I wonder whether it'd be worth also/instead supporting masking for flux?

This is working code, inserted [here in transformer_flux.py](https://github.com/huggingface/diffusers/blob/a7d53a59394d5d8367826663601b69828e9f74fc/src/diffusers/models/transformers/transformer_flux.py#L475):
```py
if joint_attention_kwargs is not None and "encoder_attention_mask" in joint_attention_kwargs and joint_attention_kwargs["encoder_attention_mask"] is not None:
    encoder_attention_mask = joint_attention_kwargs.pop("encoder_attention_mask")
    max_seq_length = encoder_hidden_states.shape[1]
    seq_length = encoder_attention_mask.sum(dim=-1)
    batch_size = encoder_attention_mask.shape[0]
    encoder_attention_mask_with_padding = encoder_attention_mask.clone()
    for i in range(batch_size):
        current_seq_len = int(seq_length[i].item())
        if current_seq_len < max_seq_length:
            available_padding = max_seq_length - current_seq_len
            tokens_to_unmask = min(1, available_padding) # unmask one of the padding tokens
            encoder_attention_mask_with_padding[i, current_seq_len : current_seq_len + tokens_to_unmask] = 1
    attention_mask = torch.cat(
        [
            encoder_attention_mask_with_padding,
            torch.ones([hidden_states.shape[0], hidden_states.shape[1]], device=encoder_attention_mask.device),
        ],
        dim=1,
    )
    attention_mask = attention_mask.float().T @ attention_mask.float()
    attention_mask = (
        attention_mask[None, None, ...]
        .repeat(encoder_hidden_states.shape[0], self.config.num_attention_heads, 1, 1)
        .int()
        .bool()
    )
    joint_attention_kwargs["attention_mask"] = attention_mask
```

**Comment by bghira** [comment]
can you demonstrate an example where zeroing the end of the prompt is equivalent to attention masking where the softmax scores for padding sequence is near -infinity?
